### DISCUSSION TITLE
Login using email address from different domain not possible? (OIDC, Keycloak)

### DISCUSSION THREAD

**Discussion by PatrickHuetter** [outline/outline#2641] [Self Hosting]
Hey together,

we are using outline with oidc over keycloak without problems. There is just one problem when we use a different email domain to login. E.g. Most of us use an email address like name@ourcompany.tld. That's no problem, we can login easily using keycloak. 

If someone has an account like name@gmail.com in the same keycloak (OIDC provider) database he can login with the OIDC provider/flow but gets into a redirect loop in outline. This only happens for users that don't have a name@ourcompany.tld email address but they are in the same OIDC provider/database.

E.g. for example if we have some external people joining temporary our team we don't give them company email addresses and use the private mail addresses instead.

Is this a bug or can we enable login from different domains since they must be contained in the same user database (keycloak oidc provider)? It's still the same unique/single team.

**Comment by tommoor** [comment]
Do you have the `ALLOWED_DOMAINS` env variable set? I think this is the only way this can be restricted – otherwise any email address would work just fine on the same unique team.

**Reply by PatrickHuetter** [reply]
No it's not set. I'm using this helm chart with OIDC variables and nothing more. 
https://github.com/encircle360-oss/outline-helm-chart

So the case you describe doesn't seem to work like described. 

To be clear:
I have multiple accounts in my keycloak. Some with name@ourcompany.tld and some with name@gmail.com, name@yahoo.com and name@differentdomain.tld. The accounts are exactly the same (I also created two new ones to test it). The ones that don't match @ourcompany.tld can't login in outline. They can login in keycloak oauth2 flow but after redirecting back to outline outline sends them back to keycloak. Since the user is already logged in in keycloak and has a session cookie he gets directly redirected back to outline and then the process starts again (endless loop).

It's a complete fresh install of outline, no migrations before. 
I'll do another test to be absolutely sure.

**Reply by tommoor** [reply]
It's weird because there is no path I know of that would forward immediately back to Keycloak, any error or unauthorized scenario would result in landing back on the login page. Perhaps you can post some server logs?

**Reply by hyperty** [reply]
i met the same issue, logs are here:

{"error":"The maximum number of teams has been reached","stack":"UnauthorizedError: The maximum number of teams has been reached\n    at new AuthenticationError (/opt/outline/build/server/errors.js:32:34)\n    at accountProvisioner (/opt/outline/build/server/commands/accountProvisioner.js:42:11)\n    at runMicrotasks (<anonymous>)\n    at processTicksAndRejections (internal/process/task_queues.js:95:5)\n    at async OAuth2Strategy._verify (/opt/outline/build/server/routes/auth/providers/oidc.js:93:22)","level":"error","message":"Error during authentication"}

{"error":"The maximum number of teams has been reached","stack":"UnauthorizedError: The maximum number of teams has been reached\n    at new AuthenticationError (/opt/outline/build/server/errors.js:32:34)\n    at accountProvisioner (/opt/outline/build/server/commands/accountProvisioner.js:42:11)\n    at runMicrotasks (<anonymous>)\n    at processTicksAndRejections (internal/process/task_queues.js:95:5)\n    at async OAuth2Strategy._verify (/opt/outline/build/server/routes/auth/providers/oidc.js:93:22)","level":"error","message":"Error during authentication"}

{"error":"The maximum number of teams has been reached","stack":"UnauthorizedError: The maximum number of teams has been reached\n    at new AuthenticationError (/opt/outline/build/server/errors.js:32:34)\n    at accountProvisioner (/opt/outline/build/server/commands/accountProvisioner.js:42:11)\n    at processTicksAndRejections (internal/process/task_queues.js:95:5)\n    at async OAuth2Strategy._verify (/opt/outline/build/server/routes/auth/providers/oidc.js:93:22)","level":"error","message":"Error during authentication"}

{"error":"The maximum number of teams has been reached","stack":"UnauthorizedError: The maximum number of teams has been reached\n    at new AuthenticationError (/opt/outline/build/server/errors.js:32:34)\n    at accountProvisioner (/opt/outline/build/server/commands/accountProvisioner.js:42:11)\n    at processTicksAndRejections (internal/process/task_queues.js:95:5)\n    at async OAuth2Strategy._verify (/opt/outline/build/server/routes/auth/providers/oidc.js:93:22)","level":"error","message":"Error during authentication"}

{"error":"The maximum number of teams has been reached","stack":"UnauthorizedError: The maximum number of teams has been reached\n    at new AuthenticationError (/opt/outline/build/server/errors.js:32:34)\n    at accountProvisioner (/opt/outline/build/server/commands/accountProvisioner.js:42:11)\n    at processTicksAndRejections (internal/process/task_queues.js:95:5)\n    at async OAuth2Strategy._verify (/opt/outline/build/server/routes/auth/providers/oidc.js:93:22)","level":"error","message":"Error during authentication"}

**Comment by borrinimarco74** [comment]
@PatrickHuetter May I ask which values you've filled in .env for OIDC auth over keycloak (with Google as IdP)?

I'm quite in the same situation...

Regards

**Reply by PatrickHuetter** [reply]
I'm using the following env variables:
```   
      OIDC_CLIENT_ID: "outline.mydomain.com"
      OIDC_CLIENT_SECRET: "some-secret"
      OIDC_AUTH_URI: "https://auth.mydomain.com/auth/realms/mydomain/protocol/openid-connect/auth"
      OIDC_TOKEN_URI: "https://auth.mydomain.com/auth/realms/mydomain/protocol/openid-connect/token"
      OIDC_USERINFO_URI: "https://auth.mydomain.com/auth/realms/mydomain/protocol/openid-connect/userinfo"
      OIDC_DISPLAY_NAME: "mydomain auth"
      OIDC_SCOPES: "openid profile email"
```

**Comment by hyperty** [comment]
issue can be fixed by using `ALLOWED_DOMAINS` env variable set, just like: `ALLOWED_DOMAINS=domain1.com,domain2.com`

**Comment by eug-vs** [comment]
Potential fix in #2985

**Reply by almereyda** [reply]
I can confirm that it is possible to login from another email domain, when no `ALLOWED_DOMAINS` are set in the application environment, at least via email. OIDC testing is in process.

**Comment by nickk024** [comment]
Hello! I am having this issue on a new installation of outline in Docker. It is running 0.66.2 I believe. Any idea why I would be getting this error again even if the fix has been pushed?

ERR Error during authentication | error=The maximum number of teams has been reached stack=UnauthorizedError: The maximum number of teams has been reached
    at InvalidAuthenticationError (/opt/outline/build/server/errors.js:48:34)
    at accountProvisioner (/opt/outline/build/server/commands/accountProvisioner.js:67:52)
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
    at async Strategy._verify (/opt/outline/build/server/routes/auth/providers/google.js:84:22)

**Reply by tommoor** [reply]
The fix is in [v0.67.0-pre.1](https://github.com/outline/outline/releases/tag/v0.67.0-pre.1)

**Reply by nickk024** [reply]
Hey Tom, I am still having this issue after confirming with you (as long as I am correct) that the latest docker version includes this fix. Am i doing something wrong? I am able to log in using the Google Workspace domain, but not with a gmail.com address that I want to allow access to. I have not set the ALLOWED_DOMAINS environment variable, and the only other auth methods that I am using are Slack and email magic link. Any help is greatly appreciated. Thanks!

Was thinking of setting up my own Keycloak server anyways and maybe this is motivation to do that as well.
